src/main/java/org/jenkinsci/plugins/workflow/log/BufferedBuildListener.java

```diff
@@ -0,0 +1,178 @@
+import hudson.remoting.RemoteOutputStream;
+import hudson.util.StreamTaskListener;
+import java.io.BufferedOutputStream;
+import java.io.Closeable;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.util.Timer;
+import org.jenkinsci.remoting.SerializableOnlyOverRemoting;
+
+/**
+ * Unlike {@link StreamTaskListener} this does not set {@code autoflush} on the reconstructed {@link PrintStream}.
+ * It also implements buffering on output, flushed (without ProxyOutputStream.Flush) at intervals.
+ */
+final class BufferedBuildListener implements BuildListener, Closeable, SerializableOnlyOverRemoting {
+
+    private static final Logger LOGGER = Logger.getLogger(BufferedBuildListener.class.getName());
+
+    private final OutputStream out;
+    @SuppressFBWarnings(value = "SE_BAD_FIELD", justification = "using Replacement anyway, fields here are irrelevant")
+    private final PrintStream ps;
+
+    BufferedBuildListener(OutputStream out) throws IOException {
+        this.out = out;
+        ps = new PrintStream(out, false, "UTF-8");
+    }
+    
+    @Override public PrintStream getLogger() {
+        return ps;
+    }
+    
+    @Override public void close() throws IOException {
+        ps.close();
+    }
+
+    private Object writeReplace() {
+        return new Replacement(this);
+    }
+
+    private static final class Replacement implements SerializableOnlyOverRemoting {
+
+        private static final long serialVersionUID = 1;
+
+        private final RemoteOutputStream ros;
+
+        Replacement(BufferedBuildListener cbl) {
+            this.ros = new RemoteOutputStream(cbl.out);
+        }
+
+        private Object readResolve() throws IOException {
+            return new BufferedBuildListener(new DelayBufferedOutputStream(ros));
+        }
+
+    }
+
+    private static final class DelayBufferedOutputStream extends BufferedOutputStream {
+
+        // TODO make these customizable (not trivial since this system properties would need to be loaded on the master side and then remoted)
+        private static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
+        private static final long MAX_RECURRENCE_PERIOD = 10_000; // 10s
+        private static final float RECURRENCE_PERIOD_BACKOFF = 1.05f;
+
+        private long recurrencePeriod = MIN_RECURRENCE_PERIOD;
+
+        DelayBufferedOutputStream(OutputStream out) {
+            super(new FlushProofOutputStream(out)); // default buffer size: 8Kib
+            reschedule();
+        }
+
+        private void reschedule() {
+            Timer.get().schedule(new Flush(this), recurrencePeriod, TimeUnit.MILLISECONDS);
+            recurrencePeriod = Math.min((long) (recurrencePeriod * RECURRENCE_PERIOD_BACKOFF), MAX_RECURRENCE_PERIOD);
```

Comment:
Why back the recurrencePeriod off over time unconditionally? My initial thought is to back off only if there were no writes to the buffer since the last call to `#run`.

Reply:
Oh I forgot to mention: when wrapping in `timestamps {…}`, this works fine because the buffering is applied right before transmitting over Remoting, whereas `FileMonitoringController.Watcher` still runs every 100ms and `hudson.plugins.timestamper.pipeline.GlobalDecorator` is applied to that immediately. So after the process has been running for a few minutes, you will see a 10s pause, then a big chunk of output appear in the log but with every line marked with a timestamp to ~100ms precision. That is per thread; if there are multiple threads running, you see correspondingly faster output.

(`progressiveText.jelly` displays new build log content once per second. Blue Ocean seems to do something similar up to some cutoff; once the content for a given step gets even slightly long, it just seems to give up and stop showing anything. So it is only really interesting to look at this stuff in the classic UI.)

Reply:
Ok, that reasoning makes sense to me, and it seems like a good trade-off.